### DISCUSSION TITLE
sanboot -drive 0 boots the grub fallback on rocky linux

### DISCUSSION THREAD

**Discussion by uli42** [ipxe/ipxe#1325] [General]
I have created `snponly.efi`` with an enbedded script that runs `sanboot --drive 0` only. This `snponly.efi` is only offered to PXE clients we know have already been installed. It is used to shorten boottimes by directly booting from disk when we know there's not network boot required.

We experienced some problems with this, though:

UEFI boot list looks like this (network before rocky linux from disk):
```
$ efibootmgr -v
BootCurrent: 0003
Timeout: 1 seconds
BootOrder: 0003,0000
Boot0000* Rocky Linux	HD(1,GPT,b59d5408-0d77-4581-ac8e-7b0606fd148f,0x800,0x80000)/File(\EFI\rocky\shimx64.efi)
Boot0003* Network	VenHw(b2ad3248-4f72-4950-a966-cfe5062db83a,05000000)..GO..NO............U.E.F.I.:. . . .S.L.O.T. .4. .(.2.1./.0./.3.). .P.X.E. .I.P.v.4. . .B.r.o.a.d.c.o.m. .5.7.4.5.4. .1.0./.2.5.G.b.E. .S.F.P.2.8. .4.-.p.o.r.t. .O.C.P. .E.t.h.e.r.n.e.t. .A.d.a.p.t.e.r........A....................%..=..v...............................................................Gd-.;.A..MQ..L.P.X.E. .I.P.v.4. .B.r.o.a.d.c.o.m. .5.7.4.5.4. .1.0./.2.5.G.b.E. .S.F.P.2.8. .4.-.p.o.r.t. .O.C.P. .E.t.h.e.r.n.e.t. .A.d.a.p.t.e.r........BO..NO............U.E.F.I.:. . . .S.L.O.T. .4. .(.2.1./.0./.2.). .P.X.E. .I.P.v.4. . .B.r.o.a.d.c.o.m. .5.7.4.5.4. .1.0./.2.5.G.b.E. .S.F.P.2.8. .4.-.p.o.r.t. .O.C.P. .E.t.h.e.r.n.e.t. .A.d.a.p.t.e.r........A....................%..=..v...............................................................Gd-.;.A..MQ..L.P.X.E. .I.P.v.4. .B.r.o.a.d.c.o.m. .5.7.4.5.4. .1.0./.2.5.G.b.E. .S.F.P.2.8. .4.-.p.o.r.t. .O.C.P. .E.t.h.e.r.n.e.t. .A.d.a.p.t.e.r........BO
```

Booting using the created `snponly.efi` with the embedded script works to some will boot our ROCKY8 systems, but it seems not to not the (expected) `\EFI\rocky\shimx64.efi` but `\EFI\BOOT\BOOTX64.efi` or `\EFI\BOOT\fbx64.efi` (not sure how to tell which one), leading to some weird grub boot restore facility that is described e.g. here: https://github.com/system76/firmware-open/issues/216

The problem: This restore facility will restore a boot entry for rocky linux AS THE FIRST entry in the UEFI boot list and therefore effectively disable network boot permanently. The boot entry seems to be related to `/EFI/rocky/BOOTX64.CSV`:
```
$cat /boot/efi/EFI/rocky/BOOTX64.CSV 
shimx64.efi,rocky,,This is the boot entry for rocky
```

After the restore the boot list looks like this (rocky, then network, then Rocky Linux; both rocky paths are identical):
```
BootOrder: 0001,0003,0000
Boot0000* Rocky Linux	HD(1,GPT,b59d5408-0d77-4581-ac8e-7b0606fd148f,0x800,0x80000)/File(\EFI\rocky\shimx64.efi)
Boot0001* rocky	HD(1,GPT,b59d5408-0d77-4581-ac8e-7b0606fd148f,0x800,0x80000)/File(\EFI\rocky\shimx64.efi)
Boot0003* Network	VenHw(b2ad3248-4f72-4950-a966-cfe5062db83a,05000000)..GO..NO............U.E.F.I.:. . . .S.L.O.T. .4. .(.2.1./.0./.3.). .P.X.E. .I.P.v.4. . .B.r.o.a.d.c.o.m. .5.7.4.5.4. .1.0./.2.5.G.......
```

If I add the `--filename \EFI\rocky\shimx64.efi` option to sanboot it works fine, the restore screen does not show up and nothing is restored. But as we have multiple operating systems (e.g. RHEL, centos), too, we'd rather have a generic solution and not hardcode the operating system.

As you can see in above listing the shimx64 is already listed in the boot list. Shouldn't sanboot try any path from the boot list first before trying to find a bootable drive on its own?

Or in other words: how can we tell sanboot to use the shimx64.efi without hardcoding its path?

FWIW: Here's the content of the `/boot/efi`:
```
$ find /boot/efi/ -ls
        1     16 drwx------   3  root     root        16384 Jan  1  1970 /boot/efi/
      115      4 drwx------   4  root     root         4096 Oct 23 11:38 /boot/efi/EFI
      118      4 drwx------   2  root     root         4096 Oct 23 11:52 /boot/efi/EFI/BOOT
      121     92 -rwx------   1  root     root        91016 Apr  4  2024 /boot/efi/EFI/BOOT/fbx64.efi
      122    936 -rwx------   1  root     root       956144 Apr  4  2024 /boot/efi/EFI/BOOT/BOOTX64.EFI
      123      4 drwx------   3  root     root         4096 Oct 24 14:32 /boot/efi/EFI/rocky
      132      4 drwx------   2  root     root         4096 May 24 18:59 /boot/efi/EFI/rocky/fonts
      133    836 -rwx------   1  root     root       853472 Apr  4  2024 /boot/efi/EFI/rocky/mmx64.efi
      134    936 -rwx------   1  root     root       956144 Apr  4  2024 /boot/efi/EFI/rocky/shimx64.efi
      135      4 -rwx------   1  root     root         1024 Oct 24 14:32 /boot/efi/EFI/rocky/grubenv
      136   2164 -rwx------   1  root     root      2214728 May 24 18:59 /boot/efi/EFI/rocky/grubx64.efi
      137    928 -rwx------   1  root     root       948816 Apr  4  2024 /boot/efi/EFI/rocky/shimx64-rocky.efi
      138      4 -rwx------   1  root     root          104 Apr  4  2024 /boot/efi/EFI/rocky/BOOTX64.CSV
      139      8 -rwx------   1  root     root         6681 Oct 24 14:32 /boot/efi/EFI/rocky/grub.cfg
```

**Comment by NiKiZe** [comment]
You can't. Or well, you can maybe fix the invalid bootx64.efi setup on disk... But this is not an issue with iPXE

bootx64.efi is the default filename, this should work even if your firmware has no efiboot configuration. In this case it sounds like they are abusing the system.

iPXE will give the firmware the default filename to boot when using sanboot, as you know you can override this filename.
iPXE will not check efiboot configuration to find alternative filenames.
iPXE  is first and foremost for network booting, not local booting.
The sanboot "hack" is mostly to boot known local configs.

See notes about non-default filename at https://ipxe.org/cmd/sanboot#notes

You could have sanboot --filename /rocky/shim || sanboot etc
Not nice but at least that would be one iPXE build working for all.

**Reply by uli42** [reply]
understood. The reason I am (mis)using ipxe is that I want a short file. I can achieve similar behaviour with grub2 but that one is huge compared to snponly.efi

**Comment by mcb30** [comment]
> If I add the `--filename \EFI\rocky\shimx64.efi` option to sanboot it works fine, the restore screen does not show up and nothing is restored. But as we have multiple operating systems (e.g. RHEL, centos), too, we'd rather have a generic solution and not hardcode the operating system.
> 
> As you can see in above listing the shimx64 is already listed in the boot list. Shouldn't sanboot try any path from the boot list first before trying to find a bootable drive on its own?
> 
> Or in other words: how can we tell sanboot to use the shimx64.efi without hardcoding its path?

You can't.  The boot list is stored in the local machine's UEFI variable storage (in flash), and is not part of the disk image.  The UEFI boot model implicitly relies upon the incorrect assumption that there is, in all circumstances, a 1:1 correspondence between a UEFI variable store and a set of attached disks.  This UEFI assumption is broken when performing a SAN boot, since the SAN disk is not part of the booting machine.

For example: it's perfectly possible to have a single shared SAN disk, with every machine in a cluster network booting from the same SAN disk.  (The disk can be read-only, or the SAN target can use copy-on-write to provide isolation.)

Another example is when the SAN disk represents a particular workload, and an arbitrary physical machine is directed to boot from the disk in order to run the workload on demand.

So: when performing a SAN boot, the contents of the UEFI variable storage have to be treated as irrelevant.  iPXE therefore doesn't even attempt to parse them.

Almost all operating systems will install their on-disk bootloader at the UEFI standard location (`\EFI\Boot\BOOTX64.EFI` in the EFI system partition, for x86_64).  It's vanishingly rare to see a SAN disk with multiple operating systems installed and sharing a single EFI system partition, so the use of a fixed filename doesn't cause collisions in practice.  Users can then think in terms of just booting from a particular disk, which matches the conceptual model that most people have for SAN booting.  It's unfortunate that Rocky doesn't seem to comply with this model.

As @NiKiZe mentioned, you can use some combination of `--filename`, `--extra`, and script error handling to work around this issue, at least until Rocky falls in line and uses the UEFI standard location for its bootloader.
